kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaSinglePortExactlyOnceOutputOperator.java

```diff
@@ -0,0 +1,369 @@
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (!partialWindowTuples.isEmpty() && windowId > windowDataManager.getLargestRecoveryWindow()) {
+      throw new RuntimeException("Violates Exactly once. Not all the tuples received after operator reset.");
+    }
+
+    try {
+      this.windowDataManager.save(getPartitionsAndOffsets(), operatorId, windowId);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    } catch (ExecutionException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public WindowDataManager getWindowDataManager()
+  {
+    return windowDataManager;
+  }
+
+  public void setWindowDataManager(WindowDataManager windowDataManager)
+  {
+    this.windowDataManager = windowDataManager;
+  }
+
+  private boolean doesKeyBelongsToThisInstance(Integer operatorId, String key)
+  {
+    String[] split = key.split(KEY_SEPARATOR);
+
+    if (split.length != 2) {
+      return false;
+    }
+
+    if ((Integer.parseInt(split[1]) == operatorId) && (split[0].equals(appName))) {
+      return true;
+    }
+
+    return false;
+  }
+
+  private boolean alreadyInKafka(T message)
+  {
+    if ( windowId <= windowDataManager.getLargestRecoveryWindow() ) {
+      return true;
+    }
+
+    if (partialWindowTuples.containsKey(message)) {
+
+      Integer val = partialWindowTuples.get(message);
+
+      if ( val == 0 ) {
+        return false;
+      } else if ( val == 1 ) {
+        partialWindowTuples.remove(message);
+      } else {
+        partialWindowTuples.put(message, val - 1);
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  private Map<Integer,Long> getPartitionsAndOffsets() throws ExecutionException, InterruptedException
+  {
+    List<PartitionInfo> partitionInfoList = consumer.partitionsFor(getTopic());
+    List<TopicPartition> topicPartitionList = new java.util.ArrayList<>();
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+      topicPartitionList.add(new TopicPartition(getTopic(), partitionInfo.partition()) );
+    }
+
+    Map<Integer,Long> parttionsAndOffset = new HashMap<>();
+    consumer.assign(topicPartitionList);
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+
+      try {
+
+        TopicPartition topicPartition = new TopicPartition(getTopic(), partitionInfo.partition());
+        consumer.seekToEnd(topicPartition);
+        parttionsAndOffset.put(partitionInfo.partition(), consumer.position(topicPartition));
+
+      } catch (Exception ex) {
+        throw new RuntimeException(ex);
+      }
+    }
+
+    return parttionsAndOffset;
+  }
+
+  private void rebuildPartialWindow()
+  {
+    logger.info("Rebuild the partial window after " + windowDataManager.getLargestRecoveryWindow());
+
+    Map<Integer,Long> storedOffsets = getStoredOffsets();
+    Map<Integer,Long> currentOffsets = getCurrentOffsets();
+
+    if (storedOffsets == null || currentOffsets == null) {
```

Comment:
storedOffsets can be null if operators fails before the first checkpoint. And in this case the stored offsets should be 0 so the operator can load messages from beginning